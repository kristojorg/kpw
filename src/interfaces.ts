import { State } from "opds-web-client/lib/state";
import { CollectionState } from "opds-web-client/lib/reducers/collection";
import { CollectionData, BookData } from "opds-web-client/lib/interfaces";

export interface PathFor {
  (collectionUrl?: string, bookUrl?: string): string;
}

export interface ComplaintData {
  type: string;
  detail?: string;
}

export interface Link {
  href: string;
  rel?: string;
  title?: string;
  type?: string;
  role?: string;
}

export type LibraryLinks = {
  helpWebsite?: Link;
  helpEmail?: Link;
  libraryWebsite?: Link;
  tos?: Link;
  about?: Link;
  privacyPolicy?: Link;
  registration?: Link;
};

export interface LibraryData {
  id?: string;
  onlyLibrary?: boolean;
  catalogUrl: string;
  catalogName: string;
  logoUrl?: string;
  colors?: {
    background?: string;
    foreground?: string;
  };
  headerLinks?: Link[];
  cssLinks?: Link[];
  libraryLinks: LibraryLinks;
}

export interface PreloadedData {
  library: LibraryData;
  shortenUrls: boolean;
  initialState?: State;
  helmetContext?: any;
}
declare global {
  interface Window {
    // Our custom global holding data loaded by the server
    __PRELOADED_DATA__: PreloadedData;
  }
}

export interface WebpackAssets {
  CirculationPatronWeb: string[];
}

// our CPW useReducer states
export type RecommendationsState = CollectionState;
export { ComplaintsState } from "./hooks/useComplaints/reducer";

export { Theme, ButtonVariants, TextVariants } from "./theme";
// helper for theme variant prop types
export type VariantProp<VType> = Exclude<keyof VType, keyof {}>;

export type SetCollectionAndBook = (
  collectionUrl: string,
  bookUrl: string
) => Promise<{
  collectionData: CollectionData;
  bookData: BookData;
}>;

type PickAndRequire<T, K extends keyof T> = { [P in K]-?: NonNullable<T[P]> };

/** Utility to make certain keys of a type required */
export type RequiredKeys<T, K extends keyof T> = Omit<T, K> &
  PickAndRequire<T, K>;
