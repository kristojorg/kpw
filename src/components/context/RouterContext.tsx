import * as React from "react";
import OPDSRouterContextProvider, {
  RouterContext
} from "opds-web-client/lib/components/context/RouterContext";
import { Router as RouterType, Location } from "opds-web-client/lib/interfaces";
import { useHistory, matchPath, useLocation } from "react-router-dom";

export const RouterProvider: React.FC = ({ children }) => {
  const history = useHistory();
  const realLocation = useLocation();

  const router: RouterType = {
    push: history.push,
    // we need to manipulate the createHref into the type
    // signature expected by history.createHref
    createHref: (location: string | Location) =>
      typeof location === "string"
        ? history.createHref({
            pathname: location
          })
        : history.createHref(location),
    /**
     * recreating the deprecated isActive with the new
     * matchPath API
     */
    isActive: (location: string | Location) =>
      typeof location === "string"
        ? !!matchPath(realLocation.pathname, {
            path: location,
            exact: false,
            strict: false
          })
        : !!matchPath(realLocation.pathname, {
            path: location.pathname,
            exact: false,
            strict: false
          })
  };
  // remove this once we update opds web
  const castChildren = children as React.ReactChild;

  return (
    <OPDSRouterContextProvider router={router}>
      {castChildren}
    </OPDSRouterContextProvider>
  );
};

export default RouterContext;
