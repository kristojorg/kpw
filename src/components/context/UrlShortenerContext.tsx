import * as React from "react";
import UrlShortener from "../../UrlShortener";

const UrlShortenerContext = React.createContext<UrlShortener | undefined>(
  undefined
);

export const UrlShortenerProvider: React.FC<{
  urlShortener: UrlShortener;
}> = ({ urlShortener, children }) => (
  <UrlShortenerContext.Provider value={urlShortener}>
    {children}
  </UrlShortenerContext.Provider>
);

export default function useUrlShortener() {
  const context = React.useContext(UrlShortenerContext);
  if (typeof context === "undefined") {
    throw new Error(
      "useUrlShortener must be used within a UrlShortenerProvider"
    );
  }
  return context;
}
