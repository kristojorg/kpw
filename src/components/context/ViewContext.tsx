import * as React from "react";

type ViewState = "GALLERY" | "LIST";
export const ViewContext = React.createContext<
  { view: ViewState; setView: (view: ViewState) => void } | undefined
>(undefined);

export const ViewProvider: React.FC = ({ children }) => {
  const [view, setView] = React.useState<ViewState>("GALLERY");

  return (
    <ViewContext.Provider value={{ view, setView }}>
      {children}
    </ViewContext.Provider>
  );
};

function useView() {
  const context = React.useContext(ViewContext);

  if (typeof context === "undefined") {
    throw new Error("useView must be used within a ViewProvider");
  }
  return context;
}

export default useView;
