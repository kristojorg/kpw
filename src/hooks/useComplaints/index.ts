import * as React from "react";
import complaints, { initState } from "./reducer";
import { fetchComplaintTypes, postComplaint } from "./actions";
import { useDialogState } from "reakit";
import { BookData } from "opds-web-client/lib/interfaces";

/**
 * We are using react useReducer instead of redux. The only real difference
 * here is that instead of using redux-thunk, we import the thunk actions
 * here and provide them with dispatch. These thunks take dispatch and return
 * a function to perform an action.
 *
 * eg. const doSomething = dispatch => url => fetch(url);
 *
 * when we call const boundDoSomething = doSomething(dispatch),
 * boundDoSomething is a function we can then use in our app:
 * (url) => fetch(url)
 */
export default function useComplaints(book: BookData) {
  const [state, dispatch] = React.useReducer(complaints, initState);
  const dialog = useDialogState();

  const reportUrl = getReportUrl(book.raw);

  // when the hook mounts, fetch the complaint types
  React.useEffect(() => {
    fetchComplaintTypes(dispatch)(reportUrl);
  }, [reportUrl]);

  return {
    state,
    dispatch,
    // this makes it so postComplaint just takes the data.
    postComplaint: postComplaint(dispatch)(reportUrl),
    dialog
  };
}

function getReportUrl(raw: any) {
  const reportLink = raw?.link?.find?.(
    link => link?.["$"]?.["rel"]?.["value"] === "issues"
  );

  if (!reportLink) {
    return null;
  }

  return reportLink?.["$"]?.["href"]?.["value"];
}
