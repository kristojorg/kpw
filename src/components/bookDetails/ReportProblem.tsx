/** @jsx jsx */
import { jsx, Styled } from "theme-ui";
import * as React from "react";
import Modal from "../Modal";
import useComplaints from "../../hooks/useComplaints";
import { DialogDisclosure } from "reakit";
import { BookData } from "opds-web-client/lib/interfaces";
import { TextArea } from "../TextInput";
import { useForm } from "react-hook-form";
import Button from "../Button";
import { ComplaintData } from "../../interfaces";
import LoadingIndicator from "../LoadingIndicator";
import Select, { Label } from "../Select";

const getDisplayType = (type: string) =>
  type
    .replace("http://librarysimplified.org/terms/problem/", "")
    .replace(/-/g, " ")
    .split(" ")
    .map(t => t[0].toUpperCase() + t.slice(1))
    .join(" ");

type ComplaintFormData = Required<ComplaintData>;

const ReportProblem: React.FC<{ book: BookData }> = ({ book }) => {
  const { state, dialog, dispatch, postComplaint } = useComplaints(book);
  const handleClick = () => dispatch({ type: "REPORT_PROBLEM" });

  const { register, handleSubmit, errors, reset } = useForm<
    ComplaintFormData
  >();
  const cancel = () => {
    reset();
    dispatch({ type: "REPORT_PROBLEM_CANCEL" });
  };

  const onSubmit = handleSubmit(({ type, detail }) => {
    postComplaint({ type, detail });
  });
  return (
    <React.Fragment>
      <Modal
        isVisible={state.showForm}
        dialog={dialog}
        label="Report a problem"
        hide={cancel}
        sx={{ width: "100%", maxWidth: 600 }}
      >
        {state.success ? (
          <div sx={{ display: "flex", flexDirection: "column" }}>
            <Styled.h1 sx={{ fontSize: 3, textAlign: "center" }}>
              Your problem was reported. Thank you!
            </Styled.h1>
            <Button sx={{ alignSelf: "flex-end" }} onClick={cancel}>
              Done
            </Button>
          </div>
        ) : state.isPosting ? (
          <div sx={{ display: "flex", justifyContent: "center" }}>
            <LoadingIndicator />
          </div>
        ) : (
          <form
            onSubmit={onSubmit}
            sx={{
              display: "flex",
              flexDirection: "column",
              alignItems: "flex-start",
              "&>label": {
                mt: 2,
                mb: 1
              }
            }}
          >
            <Styled.h1 sx={{ alignSelf: "center", fontSize: [3, 4] }}>
              Report a problem
            </Styled.h1>
            <Label htmlFor="complaint-type">Complaint Type</Label>
            <Select
              id="complaint-type"
              name="type"
              ref={register({ required: "Please choose a type" })}
              aria-describedby="complaint-type-error"
            >
              {state.types.map(type => (
                <option key={type} value={type}>
                  {getDisplayType(type)}
                </option>
              ))}
            </Select>
            {errors.type && (
              <span
                id="complaint-type-error"
                sx={{ color: "warn", fontStyle: "italic" }}
              >
                Error: {errors.type.message}
              </span>
            )}
            <label htmlFor="complaint-body">Details</label>
            <TextArea
              id="complaint-body"
              name="detail"
              ref={register({
                required: "Please enter details about the problem."
              })}
              sx={{ alignSelf: "stretch", maxWidth: "100%" }}
              aria-describedby="complaint-body-error"
            />
            {errors.detail && (
              <span
                id="complaint-body-error"
                sx={{ color: "warn", fontStyle: "italic" }}
              >
                Error: {errors.detail.message}
              </span>
            )}
            <div sx={{ mt: 3, "&>button": { ml: 2 }, alignSelf: "flex-end" }}>
              <Button variant="flat" onClick={cancel}>
                Cancel
              </Button>
              <Button type="submit">Submit</Button>
            </div>
          </form>
        )}
      </Modal>
      <DialogDisclosure
        {...dialog}
        onClick={handleClick}
        as={Styled.a}
        sx={{ fontStyle: "italic", mt: 2, display: "inline-block" }}
        data-testid="report-problem-link"
      >
        Report a problem
      </DialogDisclosure>
    </React.Fragment>
  );
};

export default ReportProblem;
