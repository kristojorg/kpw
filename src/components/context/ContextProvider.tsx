import * as React from "react";
import { PathFor, PreloadedData } from "../../interfaces";
import UrlShortener from "../../UrlShortener";
import { LibraryProvider } from "./LibraryContext";
import { UrlShortenerProvider } from "./UrlShortenerContext";
import PathForProvider from "opds-web-client/lib/components/context/PathForContext";
import { RouterProvider } from "./RouterContext";
import OPDSStore from "opds-web-client/lib/components/context/StoreContext";
import { RecommendationsProvider } from "./RecommendationsContext";
import { ActionsProvider } from "opds-web-client/lib/components/context/ActionsContext";
import { Provider as ReakitProvider } from "reakit";
import { HelmetProvider } from "react-helmet-async";
import { ViewProvider } from "./ViewContext";
import { State } from "opds-web-client/lib/state";
import { Store } from "redux";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import ActionsCreator from "opds-web-client/lib/actions";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";
import basicAuthPlugin from "../../auth/basicAuthPlugin";

type ProviderProps = PreloadedData & {
  // we allow custom store and actions
  // to be passed in for the sake of mocking during testing
  store?: Store<State>;
  actions?: ActionsCreator;
  fetcher?: DataFetcher;
};
/**
 * Combines all of the apps context provider into a single component for simplicity
 */
const AppContextProvider: React.FC<ProviderProps> = ({
  children,
  library,
  shortenUrls,
  initialState,
  helmetContext,
  store,
  actions,
  fetcher
}) => {
  const libraryId = library.id;
  const urlShortener = new UrlShortener(library.catalogUrl, shortenUrls);
  const pathFor: PathFor = (collectionUrl, bookUrl) => {
    let path = "";
    if (libraryId) {
      path += "/" + libraryId;
    }
    if (collectionUrl) {
      const preparedCollectionUrl = urlShortener.prepareCollectionUrl(
        collectionUrl
      );
      if (preparedCollectionUrl) {
        path += `/collection/${preparedCollectionUrl}`;
      }
    }
    if (bookUrl) {
      path += `/book/${urlShortener.prepareBookUrl(bookUrl)}`;
    }
    if (!path) {
      path = "/";
    }
    return path;
  };
  const computedFetcher = fetcher ?? new DataFetcher({ adapter });
  const computedActions = actions ?? new ActionsCreator(computedFetcher);

  return (
    <HelmetProvider context={helmetContext}>
      <ReakitProvider>
        <RouterProvider>
          <PathForProvider pathFor={pathFor}>
            <OPDSStore
              initialState={initialState}
              store={store}
              authPlugins={[basicAuthPlugin]}
            >
              <RecommendationsProvider>
                <ActionsProvider
                  actions={computedActions}
                  fetcher={computedFetcher}
                >
                  <LibraryProvider library={library}>
                    <UrlShortenerProvider urlShortener={urlShortener}>
                      <ViewProvider>{children}</ViewProvider>
                    </UrlShortenerProvider>
                  </LibraryProvider>
                </ActionsProvider>
              </RecommendationsProvider>
            </OPDSStore>
          </PathForProvider>
        </RouterProvider>
      </ReakitProvider>
    </HelmetProvider>
  );
};

export default AppContextProvider;
