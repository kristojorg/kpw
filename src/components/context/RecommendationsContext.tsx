import * as React from "react";
import collection, {
  initialState,
  CollectionState
} from "opds-web-client/lib/reducers/collection";
import useThunkReducer from "../../hooks/useThunkReducer";
import { useActions } from "opds-web-client/lib/components/context/ActionsContext";

/**
 * The recommendations state is somewhat confusing. It is no longer stored
 * in redux, since it proved easier to move it into react context instead
 * of trying to add it as a new reducer to the imported opds redux store.
 * However, it does import the same reducer code as an opds "collection",
 * and will dispatch actions from the opds ActionCreator, but with its own
 * dispatch method.
 *
 * In the future, we should look to unify this all into a single redux
 * implementation, possibly by having opds export its reducers and actions
 * and then letting consuming apps build their own store and add custom
 * reducers as necessary.
 */

export const RecommendationsStateContext = React.createContext<
  CollectionState | undefined
>(undefined);
export const RecommendationsDispatchContext = React.createContext<
  React.Dispatch<any> | undefined
>(undefined);

export const RecommendationsProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useThunkReducer(collection, initialState);

  return (
    <RecommendationsStateContext.Provider value={state}>
      <RecommendationsDispatchContext.Provider value={dispatch}>
        {children}
      </RecommendationsDispatchContext.Provider>
    </RecommendationsStateContext.Provider>
  );
};

function useRecommendationsState() {
  const state = React.useContext(RecommendationsStateContext);
  const dispatch = React.useContext(RecommendationsDispatchContext);
  const { actions } = useActions();

  if (typeof state === "undefined" || typeof dispatch === "undefined") {
    throw new Error(
      "useRecommendationsState must be used within a RecommendationsContextProvider"
    );
  }
  // possibly bind all the actions with dispatch here?
  return {
    recommendationsState: state,
    recommendationsDispatch: dispatch,
    recommendationsActions: actions
  };
}

export default useRecommendationsState;
