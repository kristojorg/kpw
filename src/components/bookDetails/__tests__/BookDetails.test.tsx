import * as React from "react";
import { render, fixtures, actions, wait } from "../../../test-utils";
import merge from "deepmerge";
import { BookDetails } from "../index";
import { State } from "opds-web-client/lib/state";
import { BookData, CollectionData } from "opds-web-client/lib/interfaces";
import * as complaintActions from "../../../hooks/useComplaints/actions";
import { RecommendationsStateContext } from "../../context/RecommendationsContext";
import userEvent from "@testing-library/user-event";
import ReportProblem from "../ReportProblem";

const mockSetCollectionAndBook = jest.fn().mockResolvedValue({});

/**
 * We need to mock actions.fetchCollection so that
 * ReportProblem doesn't actually fetch anything
 */
jest
  .spyOn(actions, "fetchCollection")
  .mockImplementation(() => () => Promise.resolve({} as CollectionData));

const makeStateWithBook = (book: BookData = fixtures.book): State =>
  merge<State>(fixtures.initialState, {
    book: {
      isFetching: false,
      url: "/book",
      error: null,
      data: book
    }
  });

describe("book details page", () => {
  test("shows loading state", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: merge<State>(fixtures.initialState, {
          book: {
            isFetching: true,
            url: "/book",
            error: null,
            data: null
          }
        })
      }
    );
    expect(node.getByText("Loading...")).toBeInTheDocument();
  });

  test("handles error state", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: merge<State>(fixtures.initialState, {
          book: {
            isFetching: false,
            url: "/book",
            error: {
              response: "Some error",
              status: 401,
              url: "/book-error-url"
            },
            data: null
          }
        })
      }
    );
    /**
     * We will snapshot test this component because it isn't complicated and should prettymuch remain the same.
     */
    expect(node.container).toMatchSnapshot();
  });

  test("shows categories", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    const categories = fixtures.book.categories?.join(", ") as string;
    expect(node.getByText(categories));
    expect(node.getByText("Categories:"));
  });

  test("doesn't show categories when there aren't any", () => {
    const bookWithoutCategories: BookData = merge(fixtures.book, {
      categories: null
    });
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithoutCategories)
      }
    );
    expect(node.queryByText("Categories:")).toBeFalsy();
  });

  test("shows publisher", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    const publisher = fixtures.book.publisher as string;

    expect(node.getByText(publisher)).toBeInTheDocument();
    expect(node.getByText("Publisher:")).toBeInTheDocument();
  });

  test("shows fulfillment card", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    // there are two download buttons in the document becuase we change
    // which is displayed with media queries. But both are always rendered.
    const downloadButtons = node.getAllByText("Download");
    expect(downloadButtons).toHaveLength(2);
  });

  test("shows download requirements", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    /**
     * There are two download requirement cards in the document
     * because we display a different one depending on screen width
     */
    expect(node.getAllByText("Download Requirements:")).toHaveLength(2);
  });

  test("shows recommendation lanes", () => {
    const node = render(
      <RecommendationsStateContext.Provider
        value={{
          ...fixtures.recommendationsState
        }}
      >
        <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />
      </RecommendationsStateContext.Provider>,
      {
        initialState: makeStateWithBook()
      }
    );
    expect(node.getByText("Jane Austen")).toBeInTheDocument();
  });
});

/**
 * mock fetchComplaintTypes and postComplaint
 */
const mockBoundFetchComplaintTypes = jest
  .fn()
  .mockResolvedValue(["complaint-type"]);
const fetchComplaintTypesSpy = jest
  .spyOn(complaintActions, "fetchComplaintTypes")
  .mockReturnValue(mockBoundFetchComplaintTypes);
const postComplaintSpy = jest.spyOn(complaintActions, "postComplaint");

const bookWithReportUrl = merge<BookData>(fixtures.book, {
  raw: {
    link: [
      {
        $: {
          rel: {
            value: "issues"
          },
          href: {
            value: "/report-url"
          }
        }
      }
    ]
  }
});

describe("report problem", () => {
  test("shows report problem link", () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    const reportProblemLink = node.getByTestId("report-problem-link");
    expect(reportProblemLink).toBeInTheDocument();
  });

  test("shows form (only) when clicked", async () => {
    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    // make sure it's not visible at first
    expect(node.getByLabelText("Complaint Type")).not.toBeVisible();

    const reportProblemLink = node.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // one for the original report problem button, one for the form
    expect(node.getAllByText("Report a problem")).toHaveLength(2);
    expect(node.getByLabelText("Complaint Type")).toBeVisible();
    expect(node.getByLabelText("Details")).toBeVisible();
    expect(node.getByText("Cancel")).toBeVisible();
    expect(node.getByText("Submit")).toBeVisible();
    expect(node.getByText("Submit")).toHaveAttribute("type", "submit");
  });

  test("fetches complaint types", async () => {
    render(<BookDetails setCollectionAndBook={mockSetCollectionAndBook} />, {
      initialState: makeStateWithBook(bookWithReportUrl)
    });
    expect(mockBoundFetchComplaintTypes).toHaveBeenCalledTimes(1);
    expect(mockBoundFetchComplaintTypes).toHaveBeenCalledWith("/report-url");
  });

  test("submits", async () => {
    /**
     * Mock the fetchComplaintTypes function so that it actually
     * loads the complaintTypes into the state, so we can then
     * select them. I tried to mock the state value itself, but that
     * turned out to be quite difficult.
     */
    const complaintTypes = ["complaint-type-a", "complaint-type-b"];
    fetchComplaintTypesSpy.mockImplementationOnce(dispatch => _url => {
      dispatch({
        type: "FETCH_COMPLAINT_TYPES_SUCCESS",
        types: complaintTypes
      });
      return Promise.resolve(complaintTypes);
    });

    const mockBoundPostComplaint = jest.fn().mockResolvedValue(["some-string"]);
    postComplaintSpy.mockReturnValue(_url => mockBoundPostComplaint);

    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithReportUrl)
      }
    );
    // open the form
    const reportProblemLink = node.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // fill the form
    userEvent.selectOptions(
      node.getByLabelText("Complaint Type"),
      "complaint-type-b"
    );
    userEvent.type(node.getByLabelText("Details"), "Some issue happened.");
    // submit the form
    userEvent.click(node.getByText("Submit"));

    // actually posted the complaint
    await wait(() => expect(mockBoundPostComplaint).toHaveBeenCalledTimes(1));
    expect(mockBoundPostComplaint).toHaveBeenCalledWith({
      type: "complaint-type-b",
      detail: "Some issue happened."
    });
  });

  test("displays thank you after submitting", async () => {
    const complaintTypes = ["complaint-type-a", "complaint-type-b"];
    fetchComplaintTypesSpy.mockImplementationOnce(dispatch => _url => {
      dispatch({
        type: "FETCH_COMPLAINT_TYPES_SUCCESS",
        types: complaintTypes
      });
      return Promise.resolve(complaintTypes);
    });

    // make postComplaint dispatch a success message
    postComplaintSpy.mockImplementation(dispatch => _url => _data => {
      dispatch({ type: "POST_COMPLAINT_SUCCESS" });
      return Promise.resolve(["some string"]);
    });

    const node = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithReportUrl)
      }
    );
    // open the form
    const reportProblemLink = node.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // fill the form
    userEvent.selectOptions(
      node.getByLabelText("Complaint Type"),
      "complaint-type-b"
    );
    userEvent.type(node.getByLabelText("Details"), "Some issue happened.");
    // submit the form
    userEvent.click(node.getByText("Submit"));

    // shows thank you message
    expect(
      await node.findByText("Your problem was reported. Thank you!")
    ).toBeInTheDocument();
    expect(await node.findByText("Done")).toBeInTheDocument();

    // can close the form
    userEvent.click(node.getByText("Done"));
    expect(node.getByText("Complaint Type")).not.toBeVisible();
  });

  test("displays client error when unfilled", async () => {
    const mockBoundPostComplaint = jest.fn().mockResolvedValue(["some-string"]);
    postComplaintSpy.mockReturnValue(_url => mockBoundPostComplaint);

    const node = render(<ReportProblem book={fixtures.book} />, {
      initialState: makeStateWithBook(bookWithReportUrl)
    });
    // open the form
    const reportProblemLink = node.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // don't fill the form
    // submit the form
    userEvent.click(node.getByText("Submit"));

    expect(
      await node.findByText("Error: Please choose a type")
    ).toBeInTheDocument();
    expect(
      await node.findByText("Error: Please enter details about the problem.")
    ).toBeInTheDocument();

    expect(mockBoundPostComplaint).toHaveBeenCalledTimes(0);
  });
});
