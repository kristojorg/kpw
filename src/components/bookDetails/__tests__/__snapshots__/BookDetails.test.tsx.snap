// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book details page handles error state 1`] = `
.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  font-weight: 700;
}

.emotion-2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-primary,#0467a6);
  color: white;
  fill: white;
  margin-top: 16px;
}

.emotion-2:focus,
.emotion-2:hover {
  background-color: #03588d;
}

.emotion-2:active {
  background-color: #034874;
}

.emotion-2:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<div>
  <section
    aria-label="Book details"
    class="emotion-3"
  >
    <div>
      <p>
        There was a problem fetching this book. Please refresh the page or return home.
      </p>
      <div>
        <span
          class="emotion-0"
        >
          Error Code: 
        </span>
        401
      </div>
      <div>
        <span
          class="emotion-0"
        >
          Error Message: 
        </span>
        Some error
      </div>
      <a
        class="emotion-2"
        href="/"
      >
        Return Home
      </a>
    </div>
  </section>
</div>
`;
